maskrcnn_benchmark/utils/comm.py

```diff
@@ -64,94 +58,73 @@ def _send_and_wait(r):
-    in a list, as they were obtained from each process.
-
-    This function is useful for retrieving data from multiple processes,
-    when launching the code with torch.distributed.launch
-
-    Note: this function is slow and should not be used in tight loops, i.e.,
-    do not use it in the training loop.
-
-    Arguments:
-        data: the object to be gathered from multiple processes.
-            It must be serializable
-
+    Run all_gather on arbitrary picklable data (not necessarily tensors)
+    Args:
+        data: any picklable object
     Returns:
-        result (list): a list with as many elements as there are processes,
-            where each element i in the list corresponds to the data that was
-            gathered from the process of rank i.
+        list[data]: list of data gathered from each rank
     """
-    # strategy: the main process creates a temporary directory, and communicates
-    # the location of the temporary directory to all other processes.
-    # each process will then serialize the data to the folder defined by
-    # the main process, and then the main process reads all of the serialized
-    # files and returns them in a list
-    if not torch.distributed.is_available():
-        return [data]
-    if not torch.distributed.is_initialized():
+    world_size = get_world_size()
+    if world_size == 1:
         return [data]
-    synchronize()
-    # get rank of the current process
-    rank = torch.distributed.get_rank()
-
-    # the data to communicate should be small
-    data_to_communicate = torch.empty(256, dtype=torch.uint8, device="cuda")
-    if rank == 0:
-        # manually creates a temporary directory, that needs to be cleaned
-        # afterwards
-        tmp_dir = tempfile.mkdtemp()
-        _encode(data_to_communicate, tmp_dir)
-
-    synchronize()
-    # the main process (rank=0) communicates the data to all processes
-    torch.distributed.broadcast(data_to_communicate, 0)
-
-    # get the data that was communicated
-    tmp_dir = _decode(data_to_communicate)
-
-    # each process serializes to a different file
-    file_template = "file{}.pth"
-    tmp_file = os.path.join(tmp_dir, file_template.format(rank))
-    torch.save(data, tmp_file)
-
-    # synchronize before loading the data
-    synchronize()
-
-    # only the master process returns the data
-    if rank == 0:
-        data_list = []
-        world_size = torch.distributed.get_world_size()
-        for r in range(world_size):
-            file_path = os.path.join(tmp_dir, file_template.format(r))
-            d = torch.load(file_path)
-            data_list.append(d)
-            # cleanup
-            os.remove(file_path)
-        # cleanup
-        os.rmdir(tmp_dir)
-        return data_list
+
+    # serialized to a Tensor
+    buffer = pickle.dumps(data)
+    storage = torch.ByteStorage.from_buffer(buffer)
+    tensor = torch.ByteTensor(storage).to("cuda")
```

Comment:
Actually, @ppwwyyxx this will probably be problematic for large datasets, as we will run out of memory on the GPU when trying to perform this communication.

The idea that I had was to use shared memory on the CPU and communicate the address of the shared memory, but this doesn't work on the multiple-machine case.

Reply:
I agree. But is there any way to do all-gather on CPUs (given that the dist backend was initialized with "nccl")?

Reply:
With c10d, it is now possible to have more than one dist backend at a time. So one could potentially have one nccl backend and one mpi backend?

Reply:
Yes, I am using this communication codes in other task. Because of the large size of each `data`,  I get OOM error.

Reply:
@yelantingfeng I'd recommend either:
- reverting this change locally for now
- try creating a new process group with c10d which is on the CPU, and communicate this `data` on the CPU instead

Reply:
> Yes, I am using these communication codes in other task. Because of the large size of each `data`, I get OOM error.
I got the OOM error. Have you implement the second method recommended by @fmassa ?

Reply:
> @yelantingfeng I'd recommend either:
> 
> * reverting this change locally for now
> * try creating a new process group with c10d which is on the CPU, and communicate this `data` on the CPU instead

I think we could just set a memory limit for this all_gather function. I implement this by splitting those ByteTensors into chunks. After some tests, I found my implementation can limit the total usage of memory to MiB level. Though not 100% precise limit, but I think it should be useful enough. I would be glad to send a PR if you think this is a good improvement for this repository.